zstd_stream.go

```diff
@@ -209,24 +207,28 @@ func (r *reader) Close() error {
 func (r *reader) Read(p []byte) (int, error) {
 
 	// If we already have enough bytes, return
-	if r.dstBuffer.Len() >= len(p) {
-		return r.dstBuffer.Read(p)
+	if r.decompSize-r.decompOff >= len(p) {
+		copy(p, r.decompressionBuffer[r.decompOff:])
+		r.decompOff += len(p)
+		return len(p), nil
 	}
 
-	for r.dstBuffer.Len() < len(p) {
+	copy(p, r.decompressionBuffer[r.decompOff:r.decompSize])
+	got := r.decompSize - r.decompOff
+	r.decompSize = 0
+	r.decompOff = 0
+
+	for got < len(p) {
 		// Populate src
 		src := r.compressionBuffer
 		reader := r.underlyingReader
-		if r.srcBuffer.Len() != 0 {
-			reader = io.MultiReader(&r.srcBuffer, r.underlyingReader)
-		}
-		n, err := io.ReadFull(reader, src)
-		if err == io.EOF {
-			break
-		} else if err != nil && err != io.ErrUnexpectedEOF {
-			return 0, fmt.Errorf("failed to read from underlying reader: %s", err)
+		n, err := io.ReadFull(reader, src[r.compressionLeft:])
+		if err == io.EOF && r.compressionLeft == 0 {
+			return got, io.EOF
+		} else if err != nil && err != io.EOF && err != io.ErrUnexpectedEOF {
+			return got, fmt.Errorf("failed to read from underlying reader: %s", err)
```

Comment:
— Is it correct that 'got' is returned here instead of 0? Is this a typo, or is this intentional?

In Go, it is idiomatic to return the zero value with a non-nil error (except when dealing with EOF).

Much cleaner handling of buffers though, good work! — it should give a small performance boost also 👍

Reply:
Good call, will change that to 0.